### PR TITLE
Highlight outdated feature entries on roadmap 

### PR DESCRIPTION
**Pull request by KyleJu**
Highlight outdated feature entries for two upcoming milestones through an error icon. 

- A feature is highlighted if it is  scheduled to ship in the next 2 milestones, and its accurate_as_of date is at least 4 weeks ago.  The upcoming milestones are determined by `stable_version + 1 and + 2`
- On the backend, plumb accurate_as_of information to the roadmap page.

A part of [4370](https://github.com/GoogleChrome/chromium-dashboard/issues/4370)

### UI
Screenshot 2024-10-21 4 43 30 PM

### CODE REVIEW

**Comment by KyleJu** [comment]
@jyasskin I started on the roadmap page and used an error icon to highlight the outdated feature entries. Let me know what you think!

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 1]
```diff
@@ -234,6 +236,34 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
...
+        this.stableMilestone + 1 === this.channel?.version ||
+        this.stableMilestone + 2 === this.channel?.version
+      )
+    ) {
+      return false;
+    }
+    if (!accurateAsOf) {
+      return true;
+    }
+    const accuateDate = Date.parse(accurateAsOf);
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:257]:
```suggestion
    const accurateDate = Date.parse(accurateAsOf);
```

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 2]
```diff
@@ -234,6 +236,34 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
...
+    // 4-week period.
+    const gracePeriod = 4 * 7 * 24 * 60 * 60 * 1000;
+    if (accuateDate + gracePeriod < Date.now()) {
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:260]:
```suggestion
    if (accurateDate + gracePeriod < Date.now()) {
```

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 3]
```diff
@@ -249,6 +279,16 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
           ${f.name}
         </a>
         <span class="icon_row">
+          ${this._isFeatureOutdated(f.accurate_as_of)
+            ? html`
+                <span
+                  class="tooltip"
+                  title="Feature outdated - last updated more than four weeks ago"
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:286]: It's not quite "updated". Perhaps:
```suggestion
                  title="Feature outdated - last checked for overall accuracy more than four weeks ago"
```

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 4]
```diff
@@ -249,6 +279,16 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
...
+                >
+                  <iron-icon icon="chromestatus:error" data-tooltip></iron-icon>
+                </span>
+              `
+            : nothing}
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:291]: I note that you're missing a test for this behavior.

**Review by jyasskin — COMMENTED** [review]
This looks like a reasonable UI for the roadmap. I assume you're planning for the matching UI on the feature page to link to the feature's verification page so that owners can easily clear the warning. Thanks!

**Review by jrobbins — COMMENTED** [review]
The code looks good to me, after Jeffrey's comments are resolved.

File: client-src/elements/chromedash-roadmap-milestone-card.ts (hunk 1 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card.ts:257]: Thanks!

File: client-src/elements/chromedash-roadmap-milestone-card.ts (hunk 4 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card.ts]: Done

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 5]
```diff
@@ -234,6 +238,34 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
     `;
   }
 
+  // A feature is outdated if it is scheduled to ship in the next 2 milestones,
+  // and its accurate_as_of date is at least 4 weeks ago.
+  _isFeatureOutdated(accurateAsOf, version) {
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:243]: It's not clear what version this is. Maybe
```suggestion
  _isFeatureOutdated(accurateAsOf, liveChromeVersion) {
```
And/or write a `/** @param version ... */` block for the function?

File: client-src/elements/chromedash-roadmap-milestone-card_test.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      dateText: 'was',
+      featureHeader: 'Features in this release',
+    };
+    el = await fixture<ChromedashRoadmapMilestoneCard>(
+      html`<chromedash-roadmap-milestone-card
+        .templateContent=${templateContent}
+      ></chromedash-roadmap-milestone-card>`
+    );
+    // Represents 2024-10-23.
+    el.currentDate = 1729666830595;
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card_test.ts:25]: I think you can write this more clearly as
```suggestion
    el.currentDate = (new Date('2024-10-23')).getTime();
```

File: client-src/elements/chromedash-roadmap-milestone-card_test.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    el.stableMilestone = 10;
+    await el.updateComplete;
+  });
+
+  it('can be added to the page', async () => {
+    assert.exists(el);
+    assert.instanceOf(el, ChromedashRoadmapMilestoneCard);
+  });
+
+  it('_isFeatureOutdated tests', async () => {
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card_test.ts:35]: Can you add one test each for the icon's presence and absence in appropriate situations?

**Review by jyasskin — APPROVED** [review]

File: client-src/elements/chromedash-roadmap-milestone-card_test.ts (hunk 6 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card_test.ts:25]: Thanks!

File: client-src/elements/chromedash-roadmap-milestone-card.ts (hunk 5 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card.ts]: Done!

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 8]
```diff
@@ -234,6 +238,40 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
...
 
+  /**
+   * A feature is outdated if it is scheduled to ship in the next 2 milestones,
+   * and its accurate_as_of date is at least 4 weeks ago.
+   *
+   *  @param {string} accurateAsOf The accurate_as_of date in string.
+   *  @param {number} liveChromeVersion The Chrome milestone when a feature is live.
+   *  @return {boolean}
+   */
+  _isFeatureOutdated(accurateAsOf, liveChromeVersion) {
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:249]: You can use typescript types here:
```suggestion
   *  @param accurateAsOf The accurate_as_of date as an ISO string.
   *  @param liveChromeVersion The Chrome milestone when a feature is live.
   */
  _isFeatureOutdated(accurateAsOf: string, liveChromeVersion: number): boolean {
```

File: client-src/elements/chromedash-roadmap-milestone-card.ts [hunk 9]
```diff
@@ -249,6 +287,17 @@ class ChromedashRoadmapMilestoneCard extends LitElement {
           ${f.name}
         </a>
         <span class="icon_row">
+          ${this._isFeatureOutdated(f.accurate_as_of, this.channel?.version)
+            ? html`
+                <span
+                  class="tooltip"
+                  id="oudated-icon"
```
Comment (jyasskin) [inline client-src/elements/chromedash-roadmap-milestone-card.ts:294]: Oops:

```suggestion
                  id="outdated-icon"
```

And in all the tests.

**Review by jyasskin — APPROVED** [review]

File: client-src/elements/chromedash-roadmap-milestone-card.ts (hunk 9 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card.ts:294]: oops thanks!

File: client-src/elements/chromedash-roadmap-milestone-card.ts (hunk 8 above)
Comment (KyleJu) [reply client-src/elements/chromedash-roadmap-milestone-card.ts]: Thanks!

**Pull request merged**